src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -497,14 +564,26 @@ protected virtual void CompleteTransaction(bool isCommitted)
 					// Dispose releases blocked threads by the way.
 					Dispose();
 				}
+
+				if (isSessionProcessing && !_ignoreSessionSynchronizationFailures)
+				{
+					throw new HibernateException(
+						"A synchronization timeout occurred at transaction completion: the session was still processing. " +
+							$"You may raise {Cfg.Environment.SystemTransactionCompletionLockTimeout} if it is set too low. " +
+							"It may also be a limitation of the data provider, " +
+							"like locks applied on its side while processing transaction cancellations occurring on concurrent threads, " +
+							"thus preventing the session to finish its current processing during a transaction cancellation. " +
+							$"In such case, you may enable {Cfg.Environment.IgnoreSessionSynchronizationFailuresOnSystemTransaction}, " +
+							$"and possibly lower {Cfg.Environment.SystemTransactionCompletionLockTimeout}.");
+				}
```

Comment:
In my repeated test runs, this has occurred quite a few times with PostgreSQL. Waiting longer allow the end of the session processing. So, maybe we will need to raise the transaction completion lock timeout for PostgreSQL in order to avoid some flakiness of this test.

The timeout occurring also causes some concurrency failures to become possible since transaction completion is run concurrently to the session processing, and I have seen it once. Unfortunately in such case only the NHibernate logs allows to see there was a transaction completion lock timeout, but these do not seem to be surfaced in Teamcity logs or in test run logs from Visual Studio. So, it may cause some confusion on whether there is still concurrency issues or if it was a trouble of transaction lock timeout.